src/Microsoft.Azure.EventHubs.ServiceFabricProcessor/EventProcessorService.cs

```diff
@@ -0,0 +1,488 @@
+            }
+            finally
+            {
+                if (this.userEventProcessor != null)
+                {
+                    // partitionContext is set up before processor is created, so it is available if processor is not null
+                    await this.userEventProcessor.CloseAsync(this.partitionContext, this.linkedCancellationToken.IsCancellationRequested ? CloseReason.Cancelled : CloseReason.Failure);
+                }
+                if (receiver != null)
+                {
+                    receiver.SetReceiveHandler(null);
+                    await receiver.CloseAsync();
+                }
+                if (ehclient != null)
+                {
+                    await ehclient.CloseAsync();
+                }
+                if (this.internalFatalException != null)
+                {
+                    throw this.internalFatalException;
+                }
+            }
+        }
+
+        /// <summary>
+        /// From IPartitionReceiveHandler
+        /// </summary>
+        public int MaxBatchSize { get; set; }
+
+        async Task IPartitionReceiveHandler.ProcessEventsAsync(IEnumerable<EventData> events)
+        {
+            if ((events != null) || ((events == null) && Options.InvokeProcessorAfterReceiveTimeout))
+            {
+                IEnumerable<EventData> effectiveEvents = events;
+
+                if (effectiveEvents != null)
+                {
+                    // Save position of last event
+                    IEnumerator<EventData> scanner = effectiveEvents.GetEnumerator();
+                    EventData last = null;
+                    while (scanner.MoveNext())
+                    {
+                        last = scanner.Current;
+                    }
+                    if (last != null)
+                    {
+                        this.partitionContext.SetOffsetAndSequenceNumber(last);
+                        if (this.Options.EnableReceiverRuntimeMetric)
+                        {
+                            this.partitionContext.RuntimeInformation.Update(last);
+                        }
+                    }
+                }
+                else
+                {
+                    // Client returns null on timeout, but processor expects empty enumerable.
+                    effectiveEvents = new List<EventData>();
+                }
+
+                IEventProcessor capturedEventProcessor = this.userEventProcessor;
+                if (capturedEventProcessor != null)
+                {
+                    await capturedEventProcessor.ProcessEventsAsync(this.linkedCancellationToken, this.partitionContext, effectiveEvents);
+                }
+
+                foreach (EventData ev in effectiveEvents)
+                {
+                    ev.Dispose();
+                }
+            }
+        }
+
+        Task IPartitionReceiveHandler.ProcessErrorAsync(Exception error)
```

Comment:
>ProcessErrorAsync [](start = 38, length = 17)

Do we need to bail out on non-transient errors? This is different than EPH's logic.

Reply:
EPH must stay running, so it never bails out. With SFP, Service Fabric provides that guarantee for us. On a fatal error, just let it bubble up. User can catch it and apply their own logic in that case, or pass it through to SF and let SF do the restart, possibly on another node.